x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/optimizer/Optimizer.java

```diff
@@ -1060,9 +1076,171 @@ private Expression propagate(And and) {
+                    NotEquals neq = (NotEquals) ex;
+                    if (neq.right().foldable()) {
+                        frNotEquals.add(neq);
+                    } else {
+                        exps.add(ex);
+                    }
+                } else if (ex instanceof Range) {
+                    ranges.add((Range) ex);
+                } else if (ex instanceof BinaryComparison) {
+                    BinaryComparison bc = (BinaryComparison) ex;
+                    if (bc.right().foldable()) {
+                        frInequalities.add(bc);
+                    } else {
+                        exps.add(ex);
+                    }
+                } else {
+                    exps.add(ex);
+                }
+            }
+
+            boolean updated = false; // has the expression been modified?
+
+            // evaluate the impact of each Equal over the different types of Expressions
+            for (Iterator<Equals> iterEq = frEquals.iterator(); iterEq.hasNext(); ) {
+                Equals eq = iterEq.next();
+
+                // Equals OR NotEquals
+                for (NotEquals neq : frNotEquals) {
+                    if (eq.left().semanticEquals(neq.left())) { // a = 2 OR a != ? -> TRUE
+                        return TRUE;
+                    }
+                }
+
+                Object eqValue = eq.right().fold();
+
+                // Equals OR Range
+                for (int i = 0; i < ranges.size(); i ++) { // might modify list, so use index loop
+                    Range range = ranges.get(i);
+                    if (! eq.left().semanticEquals(range.value())) {
+                        continue;
+                    }
+
+                    Integer lowerComp = range.lower().foldable() ? BinaryComparison.compare(eqValue, range.lower().fold()) : null;
+                    Integer upperComp = range.upper().foldable() ? BinaryComparison.compare(eqValue, range.upper().fold()) : null;
+
+                    if (lowerComp != null && lowerComp == 0 && !range.includeLower()) { // a = 2 OR 2 < a < ? -> 2 <= a < ?
+                        iterEq.remove(); // update range with lower equality instead
+                        ranges.set(i, new Range(range.source(), range.value(), range.lower(), true, range.upper(), range.includeUpper()));
+                        updated = true;
+                    } else if (upperComp != null && upperComp == 0 && !range.includeUpper()) { // a = 2 OR ? < a < 2 -> ? < a <= 2
+                        iterEq.remove(); // update range with upper equality instead
+                        ranges.set(i, new Range(range.source(), range.value(), range.lower(), range.includeLower(), range.upper(), true));
+                        updated = true;
+                    } else if (lowerComp != null && upperComp != null) {
+                        if (0 < lowerComp && upperComp < 0) { // a = 2 OR 1 < a < 3
+                            iterEq.remove(); // equality is superfluous
```

Comment:
It's worth reviewing the code that does removal through position (`for (int i = ...) { ranges.remove(i); } `) since it's incorrect as after a removal it skips the next item as the counter is not decremented.
Moving that to use iterator removal is better (though more verbose).

Reply:
Right, the `ranges` index-based looping doesn't do removals in the `ranges` lists, only updates at current index (for cases like `a = 2 OR 2 < a < ? -> 2 <= a < ?`); in this case, the removal is operated on the `equals` list, but that's done through the iterator.
(But let me know please if I still miss something.)

Reply:
The removal since safe since it is followed by an add the iteration is not affected but inside `CombineBinaryComparisons` there are some removals that are not compensated for 
https://github.com/elastic/elasticsearch/blob/54c1a59650833361edb85d3a7e3bdb933d9571ba/x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/optimizer/Optimizer.java#L1138
which means the next entry in the `bcs` list might skipped.
Worth reviewing (and documenting).

Reply:
Ah, I see. Sure, I'll address that in a separate issue.